addons/mail/models/discuss/discuss_channel.py

```diff
@@ -718,13 +718,17 @@ def _find_or_create_persona_for_channel(self, guest_name, timezone, country_code
         elif self.env.user._is_public():
             is_guest_known = self.env["mail.guest"]._get_guest_from_context()
             country_id = self.env["res.country"].search([("code", "=", country_code)]).id
-            guest = self.env["mail.guest"]._find_or_create_for_channel(
-                channel=self,
-                country_id=country_id,
-                name=guest_name,
-                post_joined_message=post_joined_message,
-                timezone=timezone,
-            )
+            guest = self.env["mail.guest"]._get_guest_from_context()
+            if not guest:
+                guest = self.env["mail.guest"].create(
+                    {
+                        "country_id": country_id,
+                        "lang": get_lang(self.env).code,
+                        "name": guest_name,
+                        "timezone": timezone,
+                    }
+                ).sudo(False)
+            self.with_context(guest=guest).add_members(guest_ids=guest.ids, post_joined_message=post_joined_message)
             if not is_guest_known:
                 guest._set_auth_cookie()
```

Comment:
can be moved to `if not guest` (same value)

Reply:
I feel like it's easier this way? How would you do it? 

```python
channel = self
if not guest:
    ...
    channel = channel.with_context(guest=guest)
channel.add_members(...)
```

Reply:
We'll have the same issue as in websocket if we set in on a variable
self should be updated directly with auth cookie and new context

Reply:
Ok, my linter was not happy about assigning to self, but it looks like it is a common pattern afterall.